pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	controllerName = "MemberListCluster"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash
+	// in order to improve the quality of the hash distribution, refs https://github.com/golang/groupcache/issues/29
+	defaultVirtualNodeReplicas = 50
+	// How long to wait before retrying the processing of a externalIPPool change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Default number of workers processing an externalIPPool change.
+	defaultWorkers = 4
+)
+
+// default Hash Fn is crc32.ChecksumIEEE
+var defaultHashFn func(data []byte) uint32
+
+var typeAssertionErr = fmt.Errorf("type assertion error")
+
+type ClusterInterface interface {
+	Run(stopCh <-chan struct{})
+}
+
+type clusterNodeEventHandler func(objName string)
+
+type Cluster struct {
+	bindPort int
+	// nodeConfig contains nodeName and ipAddr which would be used
+	localNodeIP net.IP
+	// Name of local node. Node name must be unique in the cluster.
+	nodeName string
+
+	mList *memberlist.Memberlist
+	// consistentHash hold the consistenthash.Map, when a node join cluster, use method Add() to add some keys to the hash.
+	// when a node leave the cluster, the consistenthash.Map should update
+	consistentHashMap     map[string]*consistenthash.Map
+	consistentHashRWMutex sync.RWMutex
+	// nodeEventsCh, the Node join/leave events will be notified via it.
+	nodeEventsCh chan memberlist.NodeEvent
+
+	// clusterNodeEventHandlers contains eventHandler of cluster node status update(leave of join cluster)
+	// the func that should implement when a node join or leave the cluster
+	// for example, when a new node join the cluster,
+	// each node should recalculate that if it should still hold the existing egresses owned by itself
+	// when a node leave cluster, each node may get some new egresses from the left node
+	clusterNodeEventHandlers []clusterNodeEventHandler
+
+	nodeInformer     coreinformers.NodeInformer
+	nodeLister       corelisters.NodeLister
+	nodeListerSynced cache.InformerSynced
+
+	externalIPPoolInformer          cache.SharedIndexInformer
+	externalIPPoolLister            crdlister.ExternalIPPoolLister
+	externalIPPoolInformerHasSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
```

Comment:
please add a comment to explain what's the key in the queue.

Reply:
done